commands for adding and removing modules

simkam opened the pull request.

Ladicek commented:
Looks good on the first sight. I'll have a look at it next week again and merge, I don't anticipate any big changes. In the meantime, could you please get rid of the `@author` javadoc tags? You can extract this info from Git already :-)

Ladicek commented on testsuite/standalone/src/test/java/org/wildfly/extras/creaper/commands/modules/AddRemoveModuleTest.java:
```diff
@@ -0,0 +1,120 @@
+        File asRoot = new File(getPathToAs());
+
+        File module = new File(asRoot, "modules" + File.separator + TEST_MODULE_NAME.replaceAll("\\.", File.separator));
+        assertTrue("Module " + module.getAbsolutePath() + " doesn't exist on path", module.exists());
```
I always found this style of assertion failure messages confusing. I'd prefer the "should" style; in this case, `Module ... should exist on path`.

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/modules/AddModule.java:
```diff
@@ -0,0 +1,170 @@
+
+        Joiner joiner = Joiner.on(',');
+        if (!resources.isEmpty()) {
+            cmd.append(" --resources=" + joiner.join(resources));
```
What happens if the path to the resource contains spaces (and/or quotes)? See also the `PatchingConversions` class. There should probably be some kind of a builder for local CLI operations that would encapsulate that, but for the time being, you have to solve that manually.

Ladicek commented:
BTW, please also rebase on master.

simkam commented:
* fixed resource-delimiter
* assertion failure messages changed to 'should' style
* spaces in resources paths are now escaped 
  * I think that quotes in path are so rare that is no need to handle them specially, also `Builder` accepts resource as `File` only, so I don't expect that anybody mistakenly adds path encapsulated in quotes 
* added path with spaces to test

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/modules/AddModule.java:
```diff
@@ -0,0 +1,171 @@
+
+        Joiner resourcesJoiner = Joiner.on(resourceDelimiter);
+        if (!resources.isEmpty()) {
+            cmd.append(" --resources=" + resourcesJoiner.join(resources).replaceAll(" ", "\\\\ "));
```
What if the resource delimiter is a space here? It probably can't be, but that's not documented anywhere. I mean -- I'm fine with throwing an exception if an assumption is violated, but if there are such assumptions, they should be checked (and documented).

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/modules/AddModule.java:
```diff
@@ -0,0 +1,171 @@
+    }
+
+    public static final class Builder {
+        public static final String DEFAULT_SLOT = "main";
```
I'd probably move this to the top-level `AddModule` class, because I think that a builder shouldn't hold constants related to the class it builds, but I'll leave that up to you.

I was thinking about making it `private`/`protected`/default, but I agree that it makes sense as `public`.

Ladicek commented:
Let's fix that string escaping thing, otherwise LGTM.

simkam commented:
* added validation of values in Builder
* changed `resourceDelimiter` type to `char`
* default slot constant moved `AddModule` class

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/modules/AddModule.java:
```diff
@@ -0,0 +1,197 @@
+        /**
+         * @param moduleName the name of the module to be added
+         * @param slot       specifies a slot which should be created. If this argument is not specified, "main" slot
+         *                   is assumed.slot of the new module
```
`is assumed.slot of the new module`? :-)

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/modules/RemoveModule.java:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * @param moduleName the name of the module to be removed
+     * @param slot       specifies a slot which should be removed. If this argument is not specified, "main" slot
+     *                   is assumed.slot of the new module
```
`is assumed.slot of the new module`? :-)

Ladicek commented:
Besides the two tiny issues, the test fails with EAP 6 with this exception: `org.jboss.as.cli.CommandFormatException: Unrecognized arguments: [--resource-delimiter]`. Not really sure about this... do we even need that option?

simkam commented:
updated
* fixed javadoc
* --resource-delimiter is ignored if `serverVersion` < 2.0.0  (WF8)

test still fail with EAP6, it doesn't accept paths with spaces. What to do? 

all fails with EAP6
module add --name=test --resources=/tmp/path with spaces/jboss-negotiation-extras.jar
module add --name=test --resources=/tmp/path\ with\ spaces/jboss-negotiation-extras.jar
module add --name=test --resources="/tmp/path\ with\ spaces/jboss-negotiation-extras.jar"
module add --name=test --resources="/tmp/path with spaces/jboss-negotiation-extras.jar"

Ladicek commented on commands/src/main/java/org/wildfly/extras/creaper/commands/modules/AddModule.java:
```diff
@@ -0,0 +1,204 @@
+
+        Joiner resourcesJoiner = Joiner.on(File.pathSeparatorChar);
+        // resource-delimiter was added in WF 8, WFLY-1871
+        if (ctx.serverVersion.greaterThan(ManagementVersion.VERSION_2_0_0)) {
```
Should this be `greaterThanOrEqualTo`?

simkam replied:
fixed

simkam commented:
* category `WildFlyTests` added to test
* version check fixed
